protostar/commands/test/cheatcodes/given_cheatcode.py

```diff
@@ -41,17 +41,14 @@ def given(self, **kwargs: Any) -> None:
                     f"Type {type(descriptor)} is not a valid fuzzing strategy.",
                 )
 
-            learned |= self.strategy_selector.learn(param, descriptor)
-
-        if learned:
-            raise StrategyLearnedException
-
-
-class StrategyLearnedException(BaseException):
-    """
-    An exception raised from the ``given`` cheatcode, indicating that the set of fuzzing strategies
-    has changed (fuzzer _learned_ a new strategy).
+            # Raise an error if user is trying to reassign a strategy descriptor.
+            # While this is not a big problem for us, it is a rare situation that somebody
+            # might want to do this explicitly, and more often this is just a typo bug.
+            if param in self.test_config.fuzz_declared_strategies:
```

Comment:
Possible scenario, not sure if it makes sense:
- I have 5 test cases with params `a`, `b`
- I want to have the same strategy for 4 of them, and different for one
- I use given in setup for 5 of them
- I override it for the one in case hook

Reply:
I have explicitly added the `given` cheatcode only to the `SetupCaseExecutionEnvironment` (used only for `setup_*` cases), not `SetupExecutionEnvironment` (used by both `__setup__` and `setup_*`), in order to prevent such usage patterns.

I find trying to deduplicate strategy definitions here an antipattern because the user is prone to silently break tests by trying to tinker with strategies of other seemingly unrelated tests, defined in `__setup__` out of laziness/overengineering.